docs/proposal-new-output-syntax.md

```diff
@@ -0,0 +1,725 @@
+- `foo:fail  # ERROR`
+- `foo:submit-fail  # ERROR`
+
+Finally, optional `start` doesn't really make sense. At best it would be
+equivalent to `foo:submit?`, so use that instead.
+- `foo?start  # ERROR`
+
+
+## Path Branching
+
+### Concurrent Paths 
+
+Concurrent paths can trigger off of required outputs:
+```
+a:x => b1
+a:y => b2
+b1 & b2 => c  # join
+```
+
+### Alternate Paths
+
+Alternate paths must branch off of optional outputs, and must be joined with
+OR triggers because only one side or the other will run:
+
+```
+a:x? => b1
+a:y? => b2
+b1 | b2 => c  # join
+```
+Note joining can trigger off of required outputs, because the output is
+only "required" if the task actually runs.
+
+**Failure recovery** is a common example of this:
+```
+a? => b1  # success path
+a:fail? => b2  # failure path
+b1 | b2 => c  # join
+```
+
+Note that an artificial dependency can be used to *ensure that at least one
+of the alternate paths gets taken*:
+```
+a:x? => b1
+a:y? => b2
+b1 | b2 => c  # join
+a => c # ARTIFICIAL DEPENDENCY
+```
+This forces `c` to be spawned as an unsatisfied prerequisite waiting on `b1` or
+`b2` even if neither branch runs. If the alternate paths don't actually need to
+join, `c` can be a dummy task.
+
+*See Appendix Output Completion Expressions for a possible follow-up that
+would allow us to flag `a` as incomplete here and avoid the artificial
+dependency.*
+
+## Failed Tasks
+
+No special handling of failed tasks is required.
+- if success is required, a failed task automatically gets retained in the task
+  pool as incomplete.
+- if success is optional (which would usually imply that optional failure is
+  explicitly handled), a failed task can be removed as complete
+
+No special syntax (e.g. `c:fail => NULL`) or suicide triggers (e.g. `c:fail =>
+!c`) are needed for leaf tasks either, if we don't care if they succeed or fail:
+```
+a => b => c?  # success of c is optional
+```
+
+## Family Triggers
+
+Conceivably we could try to interpret the output signifier in terms of a
+collective pseudo-output for the family as whole, but that gets really hard to
+understand.
+
+Fortunately, the output type signifier can simply be applied directly to all
+family members, for all family trigger types. That's easy to understand,
+and it makes sense because the signifier doesn't affect triggering, it only
+affects what we do with family members if the outputs in question are not
+completed (i.e. keep them in n=0 as incomplete, or forget them).
+
+```
+A:succeed-all => b
+```
+- trigger b only if all members succeed; and all members are required to
+  succeed
+
+```
+A:succeed-all? => b
+```
+- trigger b only if all members succeed; but it's OK if they fail (maybe the
+  failed path is handled elsewhere)
+
+```
+A:succeed-any => b
+```
+- trigger b if any member succeeds; but any member that runs is required to
+  succeed
+
+```
+A:succeed-any? => b
+```
+- trigger b if any member succeeds; but it's OK if any or all members fail
```

Comment:
Do you mean that `succeed` is optional for all members of the family?
If so, consider this workflow:
```
a => b1? & b2
b1? | b2 => c
```
Here `b1` may fail but `b2` is expected to succeed. `c` should run as soon as either succeed.
We wouldn't be able to use `B:succeed-any? => c` in this case (because  `succeed` would be both required and optional for `b2`).

The alternative would be to say that `:succeed-any?` doesn't imply anything about the family member outputs (same for `:succeed-all`?).

Not sure what I think about this.

Reply:
> Do you mean that succeed is optional for all members of the family?

Yes. I was thinking that the output syntax on a family would allow us to set **all** member to required or optional, and any exceptions to that would have to be handled by singling out individual members (as we already have to do with family triggers, if individual members have different dependencies).

Reply:
> any exceptions to that would have to be handled by singling out individual members

You can't do this can you? If the family trigger has made the succeed optional then you can't change it to be required (outputs can't be both required and optional).

Reply:
OK families require more thought. The problem is `succeed-all` etc. are not actually task outputs. They're trigger conditions based on some collective measure of member outputs.

Reply:
(Family proposal updated after discussion on Element chat).